inc/namespace.php

```diff
@@ -33,67 +34,145 @@ function register() {
+	 *
+	 * This allows modules to register additional documentation sets or replace the requested set.
+	 *
+	 * @param Set[]  $doc_sets Map of set ID to Set object.
+	 * @param string $set_id   The required set id.
+	 */
+	$all_sets = apply_filters( 'altis.documentation.sets', $all_sets );
+
+	if ( ! empty( $all_sets[ $set_id ] ) ) {
+		return $all_sets[ $set_id ];
 	}
 
-	$modules = Module::get_all();
+	return new Set();
+}
 
-	$other_docs = dirname( __DIR__ ) . '/other-docs';
-	$docs = [
-		'welcome' => new Group( 'Welcome' ),
-		'getting-started' => new Group( 'Getting Started' ),
-		'guides' => new Group( 'Guides' ),
-	];
-	$docs['welcome']->add_page( '', parse_file( $other_docs . '/welcome.md', $other_docs ) );
-	add_docs_for_group( $docs['getting-started'], $other_docs . '/getting-started' );
-	add_docs_for_group( $docs['guides'], $other_docs . '/guides' );
+/**
+ * Filter the $all_sets array to add the developer docs Set if it doesn't yet exist.
+ *
+ * @param array $sets The array of Documentation sets.
+ *
+ * @return array
+ */
+function filter_add_dev_docs_set( array $sets ) : array {
 
-	uasort( $modules, function ( Module $a, Module $b ) : int {
-		return $a->get_title() <=> $b->get_title();
-	} );
+	if ( empty( $sets['dev-docs'] ) ) {
+		// Generate the default set.
+		$dev_set = new Set( 'Developer Documentation' );
 
-	foreach ( $modules as $id => $module ) {
-		$module_docs = generate_docs_for_module( $id, $module );
-		if ( empty( $module_docs ) ) {
-			continue;
+		$other_docs = dirname( __DIR__ ) . '/other-docs';
+
+		$welcome = new Group( 'Welcome' );
+		$welcome->add_page( '', parse_file( $other_docs . '/welcome.md', $other_docs ) );
+		$dev_set->add_group( 'welcome', $welcome );
+
+		$getting_started = new Group( 'Getting Started' );
+		add_docs_for_group( $getting_started, $other_docs . '/getting-started' );
+		$dev_set->add_group( 'getting-started', $getting_started );
+
+		$guides = new Group( 'Guides' );
+		add_docs_for_group( $guides, $other_docs . '/guides' );
+		$dev_set->add_group( 'guides', $guides );
+
+		// Add all the registered modules.
+		$modules = Module::get_all();
+		uasort( $modules, function ( Module $a, Module $b ) : int {
+			return $a->get_title() <=> $b->get_title();
+		} );
+
+		foreach ( $modules as $id => $module ) {
+			$module_docs = generate_docs_for_module( $id, $module );
+			if ( $module_docs === null ) {
+				continue;
+			}
+
+			$dev_set->add_group( $id, $module_docs );
 		}
 
-		$docs[ $id ] = $module_docs;
+		$sets['dev-docs'] = $dev_set;
 	}
 
+	return $sets;
+}
+
+/**
+ * Default to our welcome group if this is our set.
+ *
+ * @param string $group_id The current group id.
+ * @param string $set_id   The current set id.
+ *
+ * @return string The required default group id.
+ */
+function get_default_group( string $group_id, string $set_id ) : string {
+	if ( empty( $group_id ) && ( $set_id === 'dev-docs' ) ) {
+		return 'welcome';
+	}
+
+	return $group_id;
+}
```

Comment:
I'm not sure how beneficial or necessary this function is as-is. As noted where the filter for the default group is used the filter would make more sense in this function, or we could just remove this function. Right now it seems to be hardcoding the default group for the `dev-docs` set to `welcome`, even if someone were to try and filter that.

Lets avoid anything too hardcoded, so maybe default group can be a property of the set, or we just rely on the filter in the `get_current_group_id()` function. I'm thinking the first available group makes sense as the default tbh.

Reply:
I eventually moved this inside the Set class. And set ids are no longer hard coded in the generic code.